text/0000-int128.md

```diff
@@ -0,0 +1,42 @@
+- Feature Name: int128
+- Start Date: 21-02-2016
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC adds the `i128` and `u128` types to Rust. Because these types are not available on all platforms, a new target flag (`target_has_int128`) is added to allow users to check whether 128-bit integers are supported. The `i128` and `u128` are not added to the prelude, and must instead be explicitly imported with `use core::{i128, u128}`.
+
+# Motivation
+[motivation]: #motivation
+
+Some algorithms need to work with very large numbers that don't fit in 64 bits, such as certain cryptographic algorithms. One possibility would be to use a BigNum library, but these use heap allocation and tend to have high overhead. LLVM has support for very efficient 128-bit integers, which are exposed by Clang in C as the `__int128` type.
+
+# Detailed design
+[design]: #detailed-design
+
+From a quick look at Clang's source, 128-bit integers are supported on all 64-bit platforms and a few 32-bit ones (those with 64-bit registers: x32 and MIPS n32). To allow users to determine whether 128-bit integers are available, a `target_has_int128` cfg is added. The `i128` and `u128` types are only available when this flag is set.
```

Comment:
I just ran a few quick tests on LLVM bitcode with `i128` and `i192` ([test.ll](https://gist.github.com/023335c64d219a509ec4)), here are my results (LLVM 3.7.1):
- `add` and `sub` are compiled down to a sequence of add with carry instructions for both `i128` and `i192`.
- `mul` and `udiv` are compiled to a call to `__multi3` and `__udivti3` for `i128`. These functions are implemented in `compiler-rt`.
- `mul` and `udiv` generate an LLVM error for `i192`: `LLVM ERROR: Unsupported library call operation!`
- `compiler-rt` contains variations of each function for 32, 64 and 128-bit integers (`__mulsi3`, `__muldi3`, `__multi3`). However, the 128-bit versions are only present on 64-bit platforms (`#ifdef __LP64__`).

In conclusion, LLVM doesn't support many operations on arbitrary sized integers. While we could in theory support 128-bit integers on all architectures, this would require extending `compiler-rt` on 32-bit architectures.